### PR TITLE
[WIP] Candidates pubkey replacement impl.

### PR DESCRIPTION
**Pull request by EanCuznaivy**
In the case of leakage of candidates' private keys.

### CODE REVIEW

File: contract/AElf.Contracts.Treasury/TreasuryContract.cs [hunk 1]
```diff
@@ -877,5 +877,16 @@ public override Dividends GetDividends(Int64Value input)
 
             return dividends;
         }
+
+        public override Empty RecordMinerReplacement(RecordMinerReplacementInput input)
```
Comment (shiwk) [inline contract/AElf.Contracts.Treasury/TreasuryContract.cs:881]: Permission check?

File: contract/AElf.Contracts.Consensus.AEDPoS/AEDPoSContract.cs [hunk 2]
```diff
@@ -201,5 +202,30 @@ public override BytesValue GetRandomBytes(BytesValue input)
...
+        public override Empty RecordCandidateReplacement(RecordCandidateReplacementInput input)
+        {
+            Assert(Context.Sender == State.ElectionContract.Value,
+                "Only Election Contract can record candidate replacement information.");
+
+            if (!TryToGetCurrentRoundInformation(out var currentRound) ||
+                !currentRound.RealTimeMinersInformation.ContainsKey(input.OldPubkey)) return new Empty();
+
+            // If this candidate is current miner, need to modify current round information.
+            var readTimeMinerInformation = currentRound.RealTimeMinersInformation[input.OldPubkey];
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.Consensus.AEDPoS/AEDPoSContract.cs:215]: readTimeMinerInformation  -> realTimeMinerInformation

File: contract/AElf.Contracts.Election/ElectionContract_Maintainence.cs [hunk 3]
```diff
@@ -260,15 +262,101 @@ public override Empty SetTreasurySchemeIds(SetTreasurySchemeIdsInput input)
...
+            blackList.Value.Add(oldPubkeyBytes);
+            State.BlackList.Value = blackList;
+
+            return new Empty();
+        }
+
+        private void PerformReplacement(string oldPubkey, string newPubkey)
+        {
+            State.CandidateReplacementMap[newPubkey] = oldPubkey;
+            State.InitialToNewestPubkeyMap[oldPubkey] = newPubkey;
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.Election/ElectionContract_Maintainence.cs:332]: Need to remove

File: contract/AElf.Contracts.Election/ElectionContract_Maintainence.cs [hunk 4]
```diff
@@ -260,15 +262,101 @@ public override Empty SetTreasurySchemeIds(SetTreasurySchemeIdsInput input)
...
+
+            var initialPubkey = State.InitialPubkeyMap[oldPubkey] ?? oldPubkey;
+            State.InitialPubkeyMap[newPubkey] = initialPubkey;
+
+            State.InitialToNewestPubkeyMap[initialPubkey] = newPubkey;
+
+            // Notify Consensus Contract to update replacement information. (Update from old record.)
+            State.AEDPoSContract.RecordCandidateReplacement.Send(new RecordCandidateReplacementInput
+            {
+                OldPubkey = initialPubkey,
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.Election/ElectionContract_Maintainence.cs:346]: initialPubkey -> oldPubkey

**Review by shiwk — APPROVED** [review]

**Review by jason-hoopox — APPROVED** [review]

**Pull request merged**